File: test/core/BasicExample.cpp
#include "Scene.hpp"
#include "GLES2Renderer.hpp"

#include "Geometries/BoxGeometry.hpp"
#include "GLES2MaterialLib.hpp"
#include "GeometryImporter.hpp"

using namespace std;
using namespace MyEngine;

int main() {

	Scene<> scene;

	auto* camera = new PerspectiveCamera<float>(90, 800.0f / 600.0f);

	scene.add(camera);

	camera->position.set(0, 0, 20);
	camera->lookAt(scene.position);

	GLES2Renderer renderer(4);

	renderer.onResize([&camera](int width, int height) {
		camera->aspect = float(width) / float(height);
		camera->updateProjectionMatrix();
		//The renderer automatically manages its internal size. No need to set it like three.js
	});

	auto* geo = new BoxGeometry<float>(10, 10, 10, 8, 4, 2);

	IMaterial* mat = createNormalMaterial<GLES2Renderer>();

	auto* box = new Mesh<float>(geo, mat);
	box->position.set(-10, 0, 0);

	scene.add(box);

	auto* suzane = GeometryImporter::ImportAsset("suzane.obj");
	suzane->position.set(10, 0, 0);

	scene.add(suzane);

	renderer.loop([&](double delta) {
		float rotation = static_cast<float>(delta * Math::PI * 0.5f);
		box->rotateY(rotation);
		suzane->rotateY(rotation);

		//Framerate is managed internally. You can tell sdl you don't want the framerate limited if you like.
		renderer.render(scene, camera);

		return false; //Don't exit early.
	});

	return 0;
}

File: test/core/BasicExample.js
"use strict";

var lib = require('nbind').init().lib;
var Clock = require('../../lib/Clock.js');

var scene = new lib.Scene();

var renderer = new lib.GLES2Renderer(4);

var camera = new lib.PerspectiveCamera(90, 800/600, 0.1, 2000);

/**
 * Note to future implementers:
 * nbind prevents you from directly operating on native attributes!
 * For instance: camera.position.set(...) does not work!
 * Instead, it will copy camera.position into a local variable
 * and set it, then destroy it. Doens't seem to be posible to fix yet.
 */
camera.position = new lib.Vector3(0, 0, 20);
camera.lookAt(scene.position);

renderer.onResize((width, height) => {
	camera.aspect = width / height;
	camera.updateProjectionMatrix();
});

scene.add(camera);

var geo = new lib.BoxGeometry(10, 10, 10, 8, 4, 2);

var mat = lib.createGLES2NormalMaterial();

var box = new lib.Mesh(geo, mat);
box.position = new lib.Vector3(-10, 0, 0);

scene.add(box);

var suzane = lib.GeometryImporter.ImportAsset("suzane.obj");
suzane.position = new lib.Vector3(10, 0, 0);

scene.add(suzane);

var clock = new Clock();

function render() {
	var delta = clock.getDelta();
	var rot = delta * Math.PI * 0.5;
	box.rotateY(rot);
	suzane.rotateY(rot);
	renderer.render(scene, camera);
	if (!renderer.needsToClose()){
		process.nextTick(render, 1); //Delay the next render only a little.
	}
}

render();
